plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,404 @@
+{
+	m_sliceSnap.addItem("Off");
+	m_sliceSnap.addItem("1/1");
+	m_sliceSnap.addItem("1/2");
+	m_sliceSnap.addItem("1/4");
+	m_sliceSnap.addItem("1/8");
+	m_sliceSnap.addItem("1/16");
+	m_sliceSnap.addItem("1/32");
+	m_sliceSnap.setValue(0);
+}
+
+void SlicerT::playNote(NotePlayHandle* handle, sampleFrame* workingBuffer)
+{
+	if (m_originalSample.frames() <= 1) { return; }
+
+	int noteIndex = handle->key() - m_parentTrack->baseNote();
+	const fpp_t frames = handle->framesLeftForCurrentPeriod();
+	const f_cnt_t offset = handle->noteOffset();
+	const int bpm = Engine::getSong()->getTempo();
+	const float pitchRatio = 1 / std::exp2(m_parentTrack->pitchModel()->value() / 1200);
+
+	float speedRatio = static_cast<float>(m_originalBPM.value()) / bpm;
+	if (!m_enableSync.value()) { speedRatio = 1; }
+	speedRatio *= pitchRatio;
+	speedRatio *= Engine::audioEngine()->processingSampleRate() / static_cast<float>(m_originalSample.sampleRate());
+
+	float sliceStart, sliceEnd;
+	if (noteIndex == 0) // full sample at base note
+	{
+		sliceStart = 0;
+		sliceEnd = 1;
+	}
+	else if (noteIndex > 0 && noteIndex < m_slicePoints.size())
+	{
+		noteIndex -= 1;
+		sliceStart = m_slicePoints[noteIndex];
+		sliceEnd = m_slicePoints[noteIndex + 1];
+	}
+	else
+	{
+		emit isPlaying(-1, 0, 0);
+		return;
+	}
+
+	if (!handle->m_pluginData) { handle->m_pluginData = new PlaybackState(sliceStart); }
+
+	float noteDone = ((PlaybackState*)handle->m_pluginData)->noteDone();
+	float noteLeft = sliceEnd - noteDone;
+
+	if (noteLeft > 0)
+	{
+		int noteFrame = noteDone * m_originalSample.frames();
+
+		SRC_STATE* resampleState = static_cast<PlaybackState*>(handle->m_pluginData)->resamplingState();
+		SRC_DATA resampleData;
+		resampleData.data_in = (m_originalSample.data() + noteFrame)->data();
+		resampleData.data_out = (workingBuffer + offset)->data();
+		resampleData.input_frames = noteLeft * m_originalSample.frames();
+		resampleData.output_frames = frames;
+		resampleData.src_ratio = speedRatio;
+
+		src_process(resampleState, &resampleData);
+
+		float nextNoteDone = noteDone + frames * (1.0f / speedRatio) / m_originalSample.frames();
+		static_cast<PlaybackState*>(handle->m_pluginData)->setNoteDone(nextNoteDone);
+
+		// exponential fade out, applyRelease() not used since it extends the note length
+		int fadeOutFrames = m_fadeOutFrames.value() / 1000.0f * Engine::audioEngine()->processingSampleRate();
+		int noteFramesLeft = noteLeft * m_originalSample.frames() * speedRatio;
+		for (int i = 0; i < frames; i++)
+		{
+			float fadeValue = static_cast<float>(noteFramesLeft - i) / fadeOutFrames;
+			fadeValue = std::clamp(fadeValue, 0.0f, 1.0f);
+			fadeValue = cosinusInterpolate(0, 1, fadeValue);
+
+			workingBuffer[i + offset][0] *= fadeValue;
+			workingBuffer[i + offset][1] *= fadeValue;
+		}
+
+		instrumentTrack()->processAudioBuffer(workingBuffer, frames + offset, handle);
+
+		emit isPlaying(noteDone, sliceStart, sliceEnd);
+	}
+	else { emit isPlaying(-1, 0, 0); }
+}
+
```

Comment:
```suggestion

void SlicerT::deleteNotePluginData(NotePlayHandle* handle)
{
	delete static_cast<PlaybackState*>(handle->m_pluginData);
}

```
Clean up PlaybackState